app/hydrosheds_app.cc

```diff
@@ -1,35 +1,275 @@
-    return 1;
-  }
+    Coordinate p1 = this->getStartingPoint(segment);
+    Coordinate p2 = this->getEndPoint(segment); 
+    return std::sqrt(std::pow(71.500 * (p1[0] - p2[0]), 2) + std::pow(111.300 * (p1[1] - p2[1]), 2));
+}
 
-  GDALAllRegister();
+double RiverSegment::getTotalLength() const
+{
+    double total_length = 0.0;
+    for(int i = 0; i < this->get_number_of_subsegments(); i++)
+    {
+        Coordinate p1 = this->getStartingPoint(i);
+        Coordinate p2 = this->getEndPoint(i); 
+        total_length += std::sqrt(std::pow(71.5 * (p1[0] - p2[0]), 2) + std::pow(111.3 * (p1[1] - p2[1]), 2)); 
+    }
+    return total_length;
+}
 
-  GDALDataset* data;
-  data = (GDALDataset*) GDALOpenEx(argv[1], GDAL_OF_VECTOR, NULL, NULL, NULL );
+double RiverSegment::getGeologicalLength() const
+{
+    double geological_length_of_feature = feature->GetFieldAsDouble("LENGTH_KM");
+    double length_of_current_segment = this->getLength();
+    double total_length = this->getTotalLength(); 
+    return (length_of_current_segment / total_length) * geological_length_of_feature;
+}
 
-  if(data == NULL)
-  {
-      std::cerr << "Opening geodatabase failed." << std::endl;
-      return 1;
-  }
+double RiverSegment::getDischarge() const
+{
+    return feature->GetFieldAsDouble("DIS_AV_CMS");
+}
 
-  std::cout << "Layer count in GeoDataBase: " << data->GetLayerCount() << std::endl;
+Coordinate RiverSegment::getStartingPoint(int seg) const
+{
+    int count = 0;
+    Coordinate p;
+    for(int i = 0; i < segment_points.size(); i += 2)
+    {
+        if (count == seg)
+        {
+            p = segment_points[i];
+            break;
+        }
+        count += 1;
+    }
+    return p;
+}
+
+Coordinate RiverSegment::getEndPoint(int seg) const
+{
+    int count = 0;
+    Coordinate p;
+    for(int i = 1; i < segment_points.size(); i += 2)
+    {
+        if (count == seg)
+        {
+            p = segment_points[i];
+            break;
+        }
+        count += 1;
+    }
+    return p;
+}
+
+bool RiverSegment::hasDownstreamSegment() const
+{
+    if(segment == segment_points.size() - 1)
+    {
+        if(feature->GetFieldAsInteger("NEXT_DOWN") == 0 || feature->GetFieldAsInteger("ENDORHEIC") == 1)
+        {
+            return false;
+        } 
+    }
+    return true;
+}
 
-  // Extract the first layer
-  auto layer = data->GetLayer(0);
+OGRFeature* RiverSegment::search_feature(unsigned int NEXT_DOWN_ID) const
+{
+
+    RiverSegment::layer->ResetReading(); 
+    std::string query = "HYRIV_ID = " + std::to_string(NEXT_DOWN_ID);
+    layer->SetAttributeFilter(query.c_str());
+    // OGRLayer* l = data->ExecuteSQL(query.c_str(), NULL, NULL);
+    // std::cout << RiverSegment::layer->GetNextFeature()->GetFieldAsInteger("HYRIV_ID") << std::endl;
+    // std::cout << " --- SQL Test End --- " << std::endl;
+    return RiverSegment::layer->GetNextFeature(); 
+    
+    /* -- standard search -- */
```

Comment:
If the above works better unconditionally, you should remove this alternative implementation.